src/bson_mapper/test/archiver_test.cpp

```diff
@@ -0,0 +1,213 @@
+    bsoncxx::oid test_oid2;
+    bsoncxx::types::b_bool test_bool;
+    bsoncxx::types::b_date test_date;
+    bsoncxx::types::b_int32 test_int32;
+    bsoncxx::types::b_int64 test_int64;
+
+    template <class Archive>
+    void serialize(Archive& ar) {
+        ar(CEREAL_NVP(test_double), CEREAL_NVP(test_utf8), CEREAL_NVP(test_doc),
+           CEREAL_NVP(test_arr), CEREAL_NVP(test_bin), CEREAL_NVP(test_oid), CEREAL_NVP(test_oid2),
+           CEREAL_NVP(test_bool), CEREAL_NVP(test_date), CEREAL_NVP(test_int32),
+           CEREAL_NVP(test_int64));
+    }
+
+    DataC()
+        : test_double{1.4},
+          test_utf8("cxx-odm"),
+          test_bin{bsoncxx::binary_sub_type::k_binary, sizeof(int),
+                   reinterpret_cast<uint8_t*>(&someInt)},
+          test_oid2("507f1f77bcf86cd799439011"),
+          test_bool{true},
+          test_date(1465225981),
+          test_int32{214},
+          test_int64{5191872984791874} {
+    }
+
+    void resetValues() {
+        test_double.value = 0.;
+        test_utf8.value = "";
+        test_bin.size = 0;
+        test_bool.value = false;
+        test_date.value = 0;
+        test_int32.value = 0;
+        test_int64.value = 0;
+    }
+};
+
+TEST_CASE(
+    "the BSON archiver enforces the existence of name-value pairs in output archives without "
+    "segfaulting") {
+    std::ofstream os("bad_data.bson", std::ios::binary);
+    cereal::BSONOutputArchive archive(os);
+
+    BadDataA a2;
+    int i = 10;
+
+    REQUIRE_THROWS(archive(a2, i));
+}
+
+TEST_CASE("the BSON archiver successfully serializes and deserializes C++") {
+    {
+        std::ofstream os("data.bson", std::ios_base::binary);
+        cereal::BSONOutputArchive archive(os);
+
+        DataA a1;
+        DataA a2;
+        DataB b1;
+
+        std::vector<int> test_arr;
+
+        test_arr.push_back(1);
+        test_arr.push_back(2);
+        test_arr.push_back(3);
+
+        std::vector<DataA> test_obj_arr;
+
+        a1.x = 42;
+        a1.z = 3.14;
+
+        a2.x = 26;
+        a2.y = 32;
+        a2.z = 3.4;
+
+        test_obj_arr.push_back(a1);
+        test_obj_arr.push_back(a2);
+
+        b1.a = 517259871609285984;
+        b1.b = 35781926586124;
+        b1.m = a2;
+        b1.arr = test_obj_arr;
+        b1.s = "hello world!";
+        b1.tp = std::chrono::system_clock::now();
+
+        DataC c1;
+
+        int i = 10;
+
+        std::chrono::system_clock::time_point tp = std::chrono::system_clock::now();
+
+        archive(CEREAL_NVP(tp), CEREAL_NVP(test_obj_arr), CEREAL_NVP(i), a1, a2, b1, c1);
+    }
+
+    {
+        DataA a1;
+        DataA a2;
+        DataB b1;
+
+        std::vector<DataA> test_obj_arr;
+
+        DataC c1;
+        c1.resetValues();
+
+        int i;
+
+        std::chrono::system_clock::time_point tp;
+
+        std::ifstream is("data.bson");
+        cereal::BSONInputArchive archive(is);
+        archive(CEREAL_NVP(tp), CEREAL_NVP(test_obj_arr), CEREAL_NVP(i), a1, a2, b1, c1);
+
```

Comment:
If you implemented an `operator==` for DataA (e.g.), you could make a DataA (a1), set the values, save it, make an empty DataA (a2), load from the save file, then compare a1 and a2. I think that would be a good addition the to the tests you already have.